### ISSUE TITLE
Nodes fail to join cluster when specifying AMI ID - kubelet cgroup driver: "cgroupfs" is different from docker cgroup driver: "systemd"

### ISSUE THREAD

**Issue by rhino5oh** [weaveworks/eksctl#3903]
**Bug encountered using eksctl 0.54.0**

When creating a cluster using Kubernetes 1.20, Ubuntu 2004 as the AMI family, and specifying an AMI ID, the cluster creation fails. The eksctl tool times out on the final step, waiting for nodes to become ready.

Upon further inspection SSH'ing into one of the nodes and doing a `journalctl -u snap.kubelet-eks.daemon`, I see the following message: `failed to run Kubelet: misconfiguration: kubelet cgroup driver: "cgroupfs" is different from docker cgroup driver: "systemd" `

What I noticed is that, when specifying an AMI ID to use instead of letting eksctl choose, the tool will for some reason create an `/etc/docker/damon.json` file with the following contents:

```
{
  "bridge": "none",
  "exec-opts": [
    "native.cgroupdriver=systemd"
  ],
  "log-driver": "json-file",
  "log-opts": {
    "max-size": "10m",
    "max-file": "10"
  },
  "live-restore": true,
  "max-concurrent-downloads": 10
}
```

Notice how it specifies "native.cgroupdriver=systemd". This is different than the cgroup driver that Kubelet is trying to use and therefore results in a failure.

The bug appears to be that eksctl creates an `/etc/docker/damon.json` file when specifying an AMI. However, if you dont specifying an AMI and use k8s 1.20 and Ubuntu 2004, the resulting nodes do NOT have that file at all.

When specifying an AMI to make the nodes with, the tool should NOT create the `/etc/docker/damon.json` file. And if it does, it certainly should not specify a cgroup driver that differs from Kublet's cgroup driver.

This can be reproduced simply by specifying the official Ubuntu 2004 AMI for k8s 1.20 with the following YAML file to create the cluster (This example with this AMI ID is for us-east-1. You can see the official EKS AMIs [here](https://cloud-images.ubuntu.com/docs/aws/eks/)):

```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: u20-plain-official-ami
  region: us-east-1
  version: "1.20"

vpc:
  subnets:
    public:
      us-east-1a: <omitted>
      us-east-1b: <omitted>
      us-east-1c: <omitted>

nodeGroups:
  - name: standard
    securityGroups:
      withShared: true
      withLocal: true
      attachIDs: ['<omitted>']
    amiFamily: Ubuntu2004
    ami: ami-0a5ba8298614a7171
    instanceType: t3.medium
    desiredCapacity: 2
    ssh:
      allow: true
      publicKeyPath: <omitted>
```

Obviously you'll have to fill in the subnet ID's, security group ID's, and public key path that is relevant to your environment.

**Comment by Callisto13** [comment]
Thanks for reporting @rhino5oh .

We removed this for the newer bootstrap codepath, but it obviously still exists in the legacy codepath which you have triggered by setting a specific AMI id. Eksctl should have also set the cgroup driver to be systemd on the kubelet, but it looks like something went awry in that configuration.

Kuberenetes' recommendation is to have the driver set to systemd, but it has been such a hassle to configure both pieces that we decided to stop doing this. It will be the default in an upcoming version so we are going to just wait for that.

I will remove the configuration from the legacy bootstrap path. There are existing workarounds noted in other issues, you can try the following:
```yaml
nodeGroups:
- name: foo
  preBootstrapCommands: 
  - "sed -i 's/cgroupDriver:.*/cgroupDriver: systemd/' /etc/eksctl/kubelet.yaml"
```
or the inverse:
```yaml
"sed -i 's/systemd/cgroupfs' /etc/docker/damon.json"
```

**Comment by rhino5oh** [comment]
@Callisto13 Thanks for the information. Your suggestion, however, will not work. In Ubuntu 20 with Kubernetes 1.20, the snap args file for kubelet (typically found at /var/snap/kubelet-eks/39/args) contains the following: `--cgroup-driver="cgroupfs"` This file's arguments override anything in /etc/eksctl/kubelet.yaml.

I will try using your preBootstrapCommand idea and modify the snap args file instead. Should I look for the fix in #3908 in 0.54.1 perhaps?

**Comment by Callisto13** [comment]
The fix will be out in 0.56.0-rc this week. That PR simply removes anything which tries to set the cgroupdriver. Eksctl does not attempt to alter that snap file, so as long as the various defaults stay on `cgroupfs` it should be fine.

> In Ubuntu 20 with Kubernetes 1.20, the snap args file for kubelet (typically found at /var/snap/kubelet-eks/39/args) contains the following: --cgroup-driver="cgroupfs" This file's arguments override anything in /etc/eksctl/kubelet.yaml.

Ah that would explain why the kubelet config written by eksctl was (mostly) ignored. This is also an example of why we abandoned trying to set this config in the new codepath, and indeed why we decided to completely stop doing our own bootstrapping (the path you are on is deprecated and will be removed soon): changes to things on disk in EKS AMIs often happen randomly and without much fanfare 🤦‍♀️
